File: examples/dekita/.claude/hooks/bash-failure-tracker.py
#!/usr/bin/env python3
"""連続Bash失敗を検知し、シェル破損時に引き継ぎプロンプト生成を提案する。

Why:
    worktreeの自己削除等でカレントディレクトリが消えると、シェルが破損状態になり
    全てのコマンドが失敗し続ける。この状態を早期検知して回復策を提示する。

What:
    - Bash失敗を連続カウント
    - シェル破損パターン（"No such file or directory"等）を検出
    - 閾値（3回連続）超過で警告と回復オプションを提示

State:
    - writes: /tmp/claude-hooks/bash-failures.json

Remarks:
    - 警告型フック（ブロックしない、回復策を提示）
    - PostToolUse:Bashで発火
    - シェル破損パターンはre.IGNORECASEでマッチング

Changelog:
    - silenvx/dekita#237: シェル破損時の自動回復メカニズム
"""

import json
import re
import tempfile
from datetime import UTC, datetime
from pathlib import Path

from lib.execution import log_hook_execution
from lib.hook_input import get_tool_result
from lib.session import parse_hook_input

# Consecutive failure threshold for warning
FAILURE_THRESHOLD = 3

# Error patterns that indicate potential shell corruption
SHELL_CORRUPTION_PATTERNS = [
    r"No such file or directory",
    r"Unable to read current working directory",
    r"cannot access",
    r"fatal: Unable to read",
]

# Tracking file location (use TMPDIR for sandbox compatibility)
TRACKING_DIR = Path(tempfile.gettempdir()) / "claude-hooks"
TRACKING_FILE = TRACKING_DIR / "bash-failures.json"


def load_tracking_data() -> dict:
    """Load existing tracking data."""
    if TRACKING_FILE.exists():
        try:
            return json.loads(TRACKING_FILE.read_text())
        except Exception:
            pass  # Best effort - corrupted tracking data is ignored
    return {"consecutive_failures": 0, "last_errors": [], "updated_at": None}


def save_tracking_data(data: dict) -> None:
    """Save tracking data."""
    TRACKING_DIR.mkdir(parents=True, exist_ok=True)
    TRACKING_FILE.write_text(json.dumps(data, indent=2))


def is_shell_corruption_error(output: str) -> bool:
    """Check if the output indicates potential shell corruption."""
    for pattern in SHELL_CORRUPTION_PATTERNS:
        if re.search(pattern, output, re.IGNORECASE):
            return True
    return False


def main():
    """PostToolUse hook for Bash commands.

    Tracks consecutive failures and suggests handoff when threshold exceeded.
    """
    result = {"continue": True}

    try:
        input_data = parse_hook_input()
        tool_result = get_tool_result(input_data) or {}

        exit_code = tool_result.get("exit_code", 0)
        stdout = tool_result.get("stdout", "")
        stderr = tool_result.get("stderr", "")
        output = f"{stdout}\n{stderr}"

        data = load_tracking_data()

        if exit_code != 0:
            # Bash command failed
            data["consecutive_failures"] += 1
            error_summary = stderr[:200] if stderr else stdout[:200]
            data["last_errors"].append(error_summary)
            # Keep only last 5 errors
            data["last_errors"] = data["last_errors"][-5:]
            data["updated_at"] = datetime.now(UTC).isoformat()
            save_tracking_data(data)

            # Check if we've hit the threshold with shell corruption patterns
            if data["consecutive_failures"] >= FAILURE_THRESHOLD:
                if is_shell_corruption_error(output):
                    result["systemMessage"] = (
                        f"⚠️ 連続 {data['consecutive_failures']} 回のBash失敗を検知。\n"
                        "シェル破損の可能性があります。\n"
                        "【対応オプション】\n"
                        "1. 引き継ぎプロンプトを生成して別セッションで継続\n"
                        "2. メインリポジトリに移動してから再実行"
                    )
                else:
                    result["systemMessage"] = (
                        f"⚠️ 連続 {data['consecutive_failures']} 回のBash失敗。\n"
                        "回復不能な場合は引き継ぎプロンプトの生成を検討してください。"
                    )
        else:
            # Bash command succeeded - reset counter
            if data["consecutive_failures"] > 0:
                data["consecutive_failures"] = 0
                data["last_errors"] = []
                data["updated_at"] = datetime.now(UTC).isoformat()
                save_tracking_data(data)

    except Exception:
        # フック実行の失敗でClaude Codeをブロックしない
        # 追跡失敗は致命的ではなく、次の実行で回復可能
        pass

    log_hook_execution("bash-failure-tracker", "approve")
    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/bash_failure_tracker.ts
#!/usr/bin/env bun
/**
 * 連続Bash失敗を検知し、シェル破損時に引き継ぎプロンプト生成を提案する。
 *
 * Why:
 *   worktreeの自己削除等でカレントディレクトリが消えると、シェルが破損状態になり
 *   全てのコマンドが失敗し続ける。この状態を早期検知して回復策を提示する。
 *
 * What:
 *   - Bash失敗を連続カウント
 *   - シェル破損パターン（"No such file or directory"等）を検出
 *   - 閾値（3回連続）超過で警告と回復オプションを提示
 *
 * State:
 *   - writes: /tmp/claude-hooks/bash-failures.json
 *
 * Remarks:
 *   - 警告型フック（ブロックしない、回復策を提示）
 *   - PostToolUse:Bashで発火
 *   - シェル破損パターンはre.IGNORECASEでマッチング
 *
 * Changelog:
 *   - silenvx/dekita#237: シェル破損時の自動回復メカニズム
 *   - silenvx/dekita#2874: TypeScriptに移植
 */

import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { logHookExecution } from "../lib/logging";
import { getToolResultAsObject, parseHookInput } from "../lib/session";

const HOOK_NAME = "bash-failure-tracker";

// Consecutive failure threshold for warning
const FAILURE_THRESHOLD = 3;

// Error patterns that indicate potential shell corruption
const SHELL_CORRUPTION_PATTERNS = [
  /No such file or directory/i,
  /Unable to read current working directory/i,
  /cannot access/i,
  /fatal: Unable to read/i,
];

// Tracking file location (use TMPDIR for sandbox compatibility)
const TRACKING_DIR = join(tmpdir(), "claude-hooks");
const TRACKING_FILE = join(TRACKING_DIR, "bash-failures.json");

interface TrackingData {
  consecutive_failures: number;
  last_errors: string[];
  updated_at: string | null;
}

/**
 * Load existing tracking data.
 */
function loadTrackingData(): TrackingData {
  if (existsSync(TRACKING_FILE)) {
    try {
      const content = readFileSync(TRACKING_FILE, "utf-8");
      return JSON.parse(content) as TrackingData;
    } catch {
      // Best effort - corrupted tracking data is ignored
    }
  }
  return { consecutive_failures: 0, last_errors: [], updated_at: null };
}

/**
 * Save tracking data.
 */
function saveTrackingData(data: TrackingData): void {
  mkdirSync(TRACKING_DIR, { recursive: true });
  writeFileSync(TRACKING_FILE, JSON.stringify(data, null, 2));
}

/**
 * Check if the output indicates potential shell corruption.
 */
export function isShellCorruptionError(output: string): boolean {
  for (const pattern of SHELL_CORRUPTION_PATTERNS) {
    if (pattern.test(output)) {
      return true;
    }
  }
  return false;
}

async function main(): Promise<void> {
  const result: { continue: boolean; systemMessage?: string } = { continue: true };
  let sessionId: string | undefined;

  try {
    const inputData = await parseHookInput();
    sessionId = inputData.session_id;
    const toolResult = getToolResultAsObject(inputData);

    // exit_code不在時はundefinedとして扱い、失敗として検知する（安全側に倒す）
    const exitCode = typeof toolResult.exit_code === "number" ? toolResult.exit_code : undefined;
    const stdout = typeof toolResult.stdout === "string" ? toolResult.stdout : "";
    const stderr = typeof toolResult.stderr === "string" ? toolResult.stderr : "";
    const output = `${stdout}\n${stderr}`;

    const data = loadTrackingData();

    // exit_code不在またはexit_code !== 0 を失敗として扱う
    if (exitCode === undefined || exitCode !== 0) {
      // Bash command failed
      data.consecutive_failures += 1;
      const errorSummary = stderr ? stderr.slice(0, 200) : stdout.slice(0, 200);
      data.last_errors.push(errorSummary);
      // Keep only last 5 errors
      data.last_errors = data.last_errors.slice(-5);
      data.updated_at = new Date().toISOString();
      saveTrackingData(data);

      // Check if we've hit the threshold with shell corruption patterns
      if (data.consecutive_failures >= FAILURE_THRESHOLD) {
        if (isShellCorruptionError(output)) {
          result.systemMessage = `⚠️ 連続 ${data.consecutive_failures} 回のBash失敗を検知。\nシェル破損の可能性があります。\n【対応オプション】\n1. 引き継ぎプロンプトを生成して別セッションで継続\n2. メインリポジトリに移動してから再実行`;
        } else {
          result.systemMessage = `⚠️ 連続 ${data.consecutive_failures} 回のBash失敗。\n回復不能な場合は引き継ぎプロンプトの生成を検討してください。`;
        }
      }
    } else {
      // Bash command succeeded - reset counter
      if (data.consecutive_failures > 0) {
        data.consecutive_failures = 0;
        data.last_errors = [];
        data.updated_at = new Date().toISOString();
        saveTrackingData(data);
      }
    }
  } catch {
    // フック実行の失敗でClaude Codeをブロックしない
    // 追跡失敗は致命的ではなく、次の実行で回復可能
  }

  logHookExecution(HOOK_NAME, "approve", undefined, undefined, { sessionId });
  console.log(JSON.stringify(result));
}

if (import.meta.main) {
  main();
}
